### PR TITLE
FEAT: certificate based authenticate Key Vault client

### PR DESCRIPTION
**Pull request by stijnmoreels**
Besides client-id/client-secret; we should also provide a way to
authenticate the Key Vault client with client-id/certificate.

 #10

### CODE REVIEW

**Comment by stijnmoreels** [comment]
Not sure if this should be included in the integration test? I see that the 'managed service identity' impelemtnation isn't used anywhere?

**Comment by tomkerkhove** [comment]
We should definitely have an integration test for this scenario. The reason why we don't have one for Managed Service Identity is that this is handled via the Azure infrastructure which we cannot use locally unless we use hacks.

Tracking lack of tests for MSI in #60

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        private readonly string _applicationId;
+        private readonly X509Certificate2 _certificate;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CertificateBasedAuthentication"/> class.
+        /// </summary>
+        /// <param name="applicationId">The identifier of the application requesting the authentication token.</param>
+        /// <param name="certificate">The certificate that is used as credential.</param>
+        public CertificateBasedAuthentication(string applicationId, X509Certificate2 certificate)
```
Comment (tomkerkhove) [inline src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: We use `clientId` in the other authentication option, can you please align?
https://github.com/arcus-azure/arcus.security/blob/master/src/Arcus.Security.Providers.AzureKeyVault/Authentication/ServicePrincipalAuthenticator.cs#L25

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            _certificate = certificate;
+        }
+
+        /// <summary>
+        ///     Authenticates with Azure Key Vault
+        /// </summary>
+        /// <returns>A <see cref="IKeyVaultClient" /> client to use for interaction with the vault</returns>
+        public Task<IKeyVaultClient> Authenticate()
+        {
+            IKeyVaultClient client = new KeyVaultClient((async (authority, resource, scope) =>
```
Comment (tomkerkhove) [inline src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: We don't tend to do inline stuff like this, please move to a dedicated method similar to our other authentication options: https://github.com/arcus-azure/arcus.security/blob/master/src/Arcus.Security.Providers.AzureKeyVault/Authentication/ServicePrincipalAuthenticator.cs#L54

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            {
+                var authenticationContext = new AuthenticationContext(authority);
```
Comment (tomkerkhove) [inline src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: Please see if we can refactor to share some code with ServicePrincipalAuthenticator

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs (hunk 1 above)
Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: Azure docs talk about Application ID; but OK to allgin.

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs (hunk 3 above)
Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: Hmm, only the 

```csharp
var authenticationContext = new AuthenticationContext(authority);
```

is a duplicate. Doesn't seem right to refactor this out.

File: src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs (hunk 1 above)
Comment (tomkerkhove) [reply src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: Then align the other part ;)

Comment (stijnmoreels) [reply src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: :) OK

Comment (tomkerkhove) [reply src/Arcus.Security.Providers.AzureKeyVault/Authentication/CertificateBasedAuthentication.cs]: That's a fair point, tracking this in #61 to align this

File: docs/auth/azure-key-vault.md [hunk 4]
```diff
@@ -30,4 +30,16 @@ var vaultConfiguration = new KeyVaultConfiguration(keyVaultUri);
 var keyVaultSecretProvider = new KeyVaultSecretProvider(vaultAuthenticator, vaultConfiguration);
 ```
 
+### Certificate Based
```
Comment (tomkerkhove) [inline docs/auth/azure-key-vault.md]: Maybe just call it `Certificate`?

Comment (stijnmoreels) [reply docs/auth/azure-key-vault.md]: Yes, that'll be better

**Comment by tomkerkhove** [comment]
Thanks!

**Pull request merged**